docs/architecture/adr-001-flush-version.md

```diff
@@ -0,0 +1,111 @@
+[PRUNING](../tree/PRUNING.md) document.
+
+These changes, while drastically improving performance under certain circumstances, introduces certain
+tradeoffs. Specifically, an application with a deterministic state-machine that commits and merkle-izes
+state via IAVL tree(s), can no longer guarantee that a state is written to disk when it is committed
+and must rely on some sort of state replay mechanism (e.g. Tendermint peer-to-peer gossip & state machine execution). While this inherently
+is not necessarily a problem, it becomes a problem under certain contexts that depend on a specific
+version existing on disk.
+
+One such example is live upgrades. Specifically, when a live upgrade occurs in some application, typically
+a new binary will be started in place of the old binary and some set of business logic will need to
+be executed in order to migrate and handle old and/or invalid state to make it compatible with the
+new version. Being that these operations must occur on the latest canonical state, we must ensure the
+latest version is committed and flushed to disk so all operators and clients see the same view when
+doing this upgrade. Hence, we need a means to manually signal to the IAVL tree that a specific
+version should be flushed to disk. In addition, we also need to ensure this version is not pruned at
+a later point in time.
+
+## Decision
+
+We propose to introduce a new API, `FlushVersion(version int64)`. This new API will attempt to manually
+flush a previously committed version to disk. It will return an error if the version does not exist
+or if the underlying flush fails. If the version is already flushed, the method performs a no-op
+and no error is returned. After the version is flushed to disk and the version is not the latest
+version, we will also remove the version from the recent (in-memory) database via `DeleteVersionFromRecent`.
+
+In order to not have a manually flushed version pruned at a later point in time
+(e.g. after an application restarts), we also introduce a new type, `CommitMetadata`, that will be
+flushed to disk for every version saved, regardless if it's flushed to disk.
+
+```go
+type PruneStrategy string
+
+const (
+  PruneStrategyNever PruneStrategy = "never"
+  PruneStrategyAuto  PruneStrategy = "auto"
+)
+
+type CommitMetadata struct {
+  Version   uint64
+  Timestamp int64
+  RootHash  []byte
+  Prune     PruneStrategy
+}
+```
+
+Currently, during `SaveVersion` we evaluate which versions to prune via `PruneRecentVersions` using
+the `Options` provided to the tree's `nodeDB`. The change we propose is now to use `CommitMetadata.Prune`
+to determine if we should actually prune a previously committed version. Under strategy `"auto"`, the
+current logic will be executed. However, if a given version is to be pruned but its strategy is `"never"`,
+the version will not be pruned. During `FlushVersion`, we automatically set the strategy to `"never"`,
+otherwise, during `SaveVersion` it's set to `"auto"`.
```

Comment:
I'm wondering if we can replace much of the current pruning logic with this. For example, we could set `Prune: yes` if the version is to be pruned on the next commit, and `Prune: no` if it is to be kept (i.e. a snapshot version). This is sort of a half-baked idea from my side, I'd like to think about it until tomorrow if that's ok.

Reply:
Yeah, my hunch when writing this was that we can simultaneously cleanup the pruning logic.  Specifically, `CommitMetadata.Prune` should be a `bool` instead of a `string`. Let me think on this some more as well, but I think we can simplify the current model.

Reply:
Yes, it might be as simple as that. There are a couple of things I'm thinking about here:

* We rely on `KeepRecent` to figure out how many versions to keep in memory. It would be nice if we could use a common mechanism to prune from memory and prune from disk.

* The SDK currently has hacks to only keep the latest `sdk.KeepEvery` version around, but all `sdk.SnapshotEvery`. We should eventually push this logic down into IAVL.

* The pruning settings may change, we should handle this in a reasonable way (see my [other comment](https://github.com/tendermint/iavl/pull/250/files#r430008815)).

* The current approach of `KeepEvery` was originally meant for state sync snapshots, but those don't need IAVL versions to be kept around, so I'm not sure if this option makes a whole lot of sense anymore.

Generally, it's always a good idea to separate policy and mechanism - I'd like to have the metadata be the mechanism and the pruning options be the policy, but I'm not entirely sure what that would look like yet.